### PR TITLE
[ISSUE #3537] Added mixpanel support

### LINKED ISSUE #3537: Add Mixpanel Library
*What*
Add the Mixpanel client library to Status

*Why*
This is the first step in collecting opt-in metrics for the purpose of understanding how the Status app is used in order to make it better. Full background is defined in [Idea 68](https://github.com/status-im/ideas/issues/68).

*Details* 
It is very important that metric collection is strictly opt-in, and anonymous. The https://github.com/status-im/ideas/issues/80 Swarm added an opt-in screen during signup, and this choice needs to be respected. 

No data at all should be sent to Mixpanel until the user explicitly opts in.

For users that opt-in, there should be a way to opt-out in settings.

Additionally, should users want to inspect the Mixpanel integration on Github the code should be clear and available. (In the Atom text editor Github created a dedicated repo for easy auditing https://github.com/atom/metrics. This is a good idea.)

### LINKED ISSUE #3545: Log events in Mixpanel
After Mixpanel is [integrated in Status](https://github.com/status-im/status-react/issues/3537) we can now start collecting basic events in order to see engagement, retention, and understand basic adoption.

[This spreadsheet has a list of basic events](https://docs.google.com/spreadsheets/d/1wV9CPA1hgTBVHg0bjT2C-ixsRSZGXmeejYluN9yuu2Y/edit#gid=0) to be instrumented.  Priority ones have an asterisk.

More info in https://github.com/status-im/ideas/issues/68

### PR DESCRIPTION
**Pull request by jeluard**
fixes #3537
fixes #3545

### Summary:

Added support for mixpanel. Some actions triggers mixpanel events

### Steps to test Mixpanel:
- Open Status
- Log in
- Opt-in for stats
- Verify events are logged in mixpanel

### Steps to test Opt-out:
- Open Status
- Log in
- Opt-OUT for stats
- Verify no events are sent to Mixpanel at all

status: ready

### CODE REVIEW

**Review by cammellos — APPROVED** [review]

**Comment by lukaszfryc** [comment]
I used CharlesProxy to check requests that are sent from mobile and then double checked events in Live View in Mixpanel (MP). Here is my initial feedback. 

Overall, the events are tracked fine. When you install the app, create new account and opt-in, the events are tracked. When not, no data is sent to MP. 
But, if you use an account that tracks data and then log out and create another account with disabled tracking, the events are still being tracked. Here are the steps I did: 
- Install the app
- Create new account and opt-in for data sharing
- Log out
- Create new account (**„Account created” is tracked but probably it should not**)
- Disable tracking for the new account
- **Events are still tracked with the same „Distinct ID”**

There are couple of things:
1. @chadyj since events tracking is enabled per account, should we keep tracking when user logged out? E.g. we track "Account created" when you create a new account after logging out
2. @chadyj do you want to use the same Distinct ID for different accounts on the same device?
3. @jeluard we should not track anything when you log into an account which had disabled the tracking in the past
 
Also, @jeluard will my tracking choice be recovered after I recover my account?

**Comment by jeluard** [comment]
@lukaszfryc Thanks! That's probably a bug where the previous tracking is not stopped when switching to a different account.

**Comment by chadyj** [comment]
Thanks @lukaszfryc!!!

> @chadyj since events tracking is enabled per account, should we keep tracking when user logged out? E.g. we track "Account created" when you create a new account after logging out

> @jeluard we should not track anything when you log into an account which had disabled the tracking in the past

This will be fixed by adding in the metrics opt-in to the recovery flow. (currently a note on the project board). The user will be able to choose opt-in settings each time they create or recover an account.

> @chadyj do you want to use the same Distinct ID for different accounts on the same device?

No. I think it makes sense to be unique for each account.

**Comment by lukaszfryc** [comment]
Cool! 

> No. I think it makes sense to be unique for each account.

Also, @jeluard please use different Distinct ID for each account as Chad's mentioned.

**Comment by jeluard** [comment]
@lukaszfryc Yes it's already the case. What you described is another issue not related with id. Don't read too much into bugs ;)

File: src/status_im/utils/handlers.cljs [hunk 1]
```diff
@@ -18,6 +18,12 @@
     (handler db params)
     db))
 
+(defn- pretty-print-event [ctx]
+  (let [[first second & _] (get-coeffect ctx :event)]
```
Comment (dmitryn) [inline src/status_im/utils/handlers.cljs]: You don't need `& _` while destructuring here.

Comment (jeluard) [reply src/status_im/utils/handlers.cljs]: @dmitryn Can you approve? I will make the change

**Review by dmitryn — APPROVED** [review]

File: src/status_im/utils/handlers.cljs (hunk 1 above)
Comment (dmitryn) [reply src/status_im/utils/handlers.cljs]: @jeluard approved

**Comment by lukaszfryc** [comment]
Looking good 👍The bugs are gone:
- When I log out and log in again to an account that has disabled data sharing the events are not tracked  
- Both Android and iOS (cc @chadyj) work fine 

Re Distinct ID, @jeluard, so in the end we will have different IDs when user logs in to another account?

Btw, there is a weird UI issue related to margins on different screens both on Android and iOS. I'm not sure how it is related to this PR. @jeluard please have a look:

**Comment by lukaszfryc** [comment]
Btw, @chadyj are we going to measure how many people enabled sharing usage data on the onboarding screen (conversion rate)?

**Comment by lukaszfryc** [comment]
Supporting recovery flow (https://github.com/status-im/status-react/issues/3632) will not be implemented in scope of this PR, right? Personally, I think we are fine without it for now. I checked it and the app does not gather any data in this case.

**Comment by jeluard** [comment]
@lukaszfryc Yes each account has a different id. Recovery flow will not be part of this PR. @chadyj do we really need it ? :)

**Comment by jeluard** [comment]
@lukaszfryc I don't think the UI issue is related to this PR. Could you check if it's in develop?

**Comment by lukaszfryc** [comment]
> Yes each account has a different id

Currently, it has the same id (Distinct ID) :) Try creating a new account and see the events in Mixpanel.

**Comment by chadyj** [comment]
> Btw, @chadyj are we going to measure how many people enabled sharing usage data on the onboarding screen (conversion rate)?

Not directly, since we won't measure those that opt-out (since they opted out!), but we can get a rough ballpark by looking at daily downloads as reported by Play/App and comparing to signups.

> Recovery flow will not be part of this PR. @chadyj do we really need it ? :)

Not for this PR, but eventually. Don't want to leave any metrics behind!  Perhaps @flexsurfer might add it to the recovery flow once he works on that.

**Comment by lukaszfryc** [comment]
Got it @chadyj 

Few more insights from my side about the events themselves. I checked [Mixpanel Events spreadsheet](https://docs.google.com/spreadsheets/d/1wV9CPA1hgTBVHg0bjT2C-ixsRSZGXmeejYluN9yuu2Y/edit#gid=0) and found out that: 

1. open-existing-chat is sent for every chat type (1-1, group, public) 
Don't we want to send additional parameter specifying the chat type?

2. Joining public chat (e.g. #status) triggers following events:
a. join-public-chat
b. create-public-chat
c. home 
d. open-existing-chat
Is c) needed?

3. When a dapp is opened:
a. open-dapp
b. open-dapp-description
c. home
d. open-selected-dapp
e. open-existing-dapp
Should c) and d) be called?

4. Looking at the I see that we will not implement following events:
a. Account created (can't track it since we don't know if user allowed for it - no account context)
b. Account recovered (same as above)
c. Sign in screen. Tap "sign in" (same thing)

@chadyj please take a look ^

**Comment by chadyj** [comment]
Looking at the recent events in live view I see that none have `app-version:`.

**Comment by chadyj** [comment]
> open-existing-chat is sent for every chat type (1-1, group, public). Don't we want to send additional parameter specifying the chat type?

Good for now. (but a good idea to consider)

> Joining public chat (e.g. #status) triggers following events:
> a. join-public-chat
> Is c) needed?

> When a dapp is opened:
> Should c) and d) be called?

Probably not. @jeluard ?

> Looking at the I see that we will not implement following events:
> a. Account created (can't track it since we don't know if user allowed for it - no account context)
> b. Account recovered (same as above)
> c. Sign in screen. Tap "sign in" (same thing)

We can't track if the user opts out, but we are tracking those events for users that opt-in.

I can see `Login`, `Account created`, and `logout` events in Mixpanel. Are they not coming through anymore?

**Comment by lukaszfryc** [comment]
> I can see Login, Account created, and logout events in Mixpanel. Are they not coming through anymore?

@jeluard @chadyj `Logout` works. But `Login` and `Account created` are not. I believe `Login` should be sent when user is logged to **existing account**. `Account created` when new account was just created (after clicking "Share usage data").

**Comment by lukaszfryc** [comment]
>I don't think the UI issue is related to this PR. Could you check if it's in develop? 

@jeluard actually, I was wrong saying that it's on develop as well. I messed up the apk I installed. Now I double checked it and develop looks good. **The issue in in our build** :O 
You can see it in my session in [TestFairy](https://app.testfairy.com/projects/4803622-status/builds/7933547/sessions/7) (from 31s).

**Comment by jeluard** [comment]
@lukaszfryc @chadyj `appVersion` should be set now.
I will look into `Login` and `Account created` events but there might be a chicken/egg issue.

**Comment by jeluard** [comment]
About extra events: events generation is generic and based on our internal actions. So this reflect what is being done currently in the code. I will look into it but removing those might require deeper changes I'd rather leave out this PR.

**Comment by lukaszfryc** [comment]
@jeluard Quick feedback about the latest version:

1. Different IDs are used for different accounts 👍 

2. Account created is logged 👍 

3. I didn't see `Login` when I logged to an existing account. See screenshot:

**Comment by jeluard** [comment]
@lukaszfryc I don't have access to this TestFairy thing

**Comment by lukaszfryc** [comment]
It's a screenshot from [Live View in Mixpanel](https://mixpanel.com/report/1541909/live/)

**Comment by jeluard** [comment]
@lukaszfryc I was referring to your previous TestFairy link related to UI issues

**Comment by lukaszfryc** [comment]
I've tested the latest version and It looks good 👍 

@chadyj the only thing we need to be aware are the events flow triggered when joining a public chat or opening a dapp. I described the issue in [this comment](https://github.com/status-im/status-react/pull/3605#issuecomment-374704080). As @jeluard said, it comes down to internal actions and how the code works. 

@chadyj I also reviewed the [mixpanel_events.edn](https://github.com/status-im/status-react/blob/9e3bc9ec446d675e6dc509128478fdf97a541f88/src/status_im/utils/mixpanel_events.edn) and I think it's quite self explanatory. How do you think? Could you include it in the wiki? Maybe we should even mention them in the onboarding screen(we could do it as a separate issue).

**Pull request merged**